provider/aws: API Gateway custom domains and base path mappings

apparentlymart opened the pull request:
This is a continuation of what @jarrodjackson started in #7708. It begins with his commit from that PR to add `aws_api_gateway_domain_name`, and then builds on it with the `aws_api_gateway_base_path_mapping` implementation from #6175 and documentation for both.

This pair of resources allow an API to be presented at a given path prefix on a custom hostname. The separation between *domain name* and *base path mapping* means that multiple APIs or API stages from a single AWS account can be presented on a single domain name.

* [x] `aws_api_gateway_domain_name` resource and acceptance tests (from #7708)
  * [x] Documentation
* [x] `aws_api_gateway_base_path_mapping` resource and acceptance tests (extracted from #6175)
  * [x] Documentation

Once complete, this should fix #7421.

radeksimko commented:
FYI: https://github.com/hashicorp/terraform/pull/6175 contains both resources. That said nobody had time to review and/or merge that PR yet and it would need to be rebased 1st either way.

The `domain` resource in this PR looks like it's _ready_ for review, so to prevent another stale WIP PR, would you mind keeping this PR focused on `aws_api_gateway_domain_name` only and then open another one for `aws_api_gateway_base_path_mapping`?

apparentlymart commented:
@radeksimko The domain name resource is pointless without the base path mapping, so my intent was to finish them both up here today (based on the work done by Jarrod and Ryan) and hopefully get these things finally merged.

I've just finished rebasing the base path mapping part of Ryan's mega-PR (and adapting it a little so it can work without the other resources he implemented) and I'm just wrapping up the docs now, so I should be able to get this out of WIP today.

apparentlymart commented:
@radeksimko I think this is ready for a review pass now. :grinning:

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping.go:
```diff
@@ -0,0 +1,130 @@
+	if err != nil {
+		if err, ok := err.(awserr.Error); ok && err.Code() == "NotFoundException" {
+			d.SetId("")
+			return nil
```
It would be IMO helpful to have a `WARN` log message here (like we have in some other resources), so that if we ever end up debugging missing resources we'd be able to track it down.

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping_test.go:
```diff
@@ -0,0 +1,137 @@
+import (
+	"fmt"
+	"testing"
+	//"time"
```
👀

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_domain_name_test.go:
```diff
@@ -0,0 +1,206 @@
+	return nil
+}
+
+// Expires on July 18, 2026
```
I thought the `tls` provider would be helpful here, but then cross-provider tests are non-trivial thing right now, so... 😐

apparentlymart replied:
Right... I remember acting as a conversation relay between @jarrodjackson and @phinze while they both agonized over this a bit, and this solution of just making a comment to explain the situation was where that discussion landed. As you noted, in theory the `tls` provider can help but we didn't want to deal with the complexity of depending on another provider here.

radeksimko replied:
Yep, I think we can avoid all this complexity and static cert is fine here.
How about using wildcard cert and randomise the subdomain?

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping_test.go:
```diff
@@ -0,0 +1,137 @@
+
+func TestAccAWSAPIGatewayBasePath_basic(t *testing.T) {
+	var conf apigateway.BasePathMapping
+	name := "test-api-gateway-domain-cert0.com"
```
I think we're going to need to put some randomisation here as the domain seems to be globally unique in AWS. Kinda like S3 bucket names. The test has failed for me:

```
TF_ACC=1 go test ./builtin/providers/aws -v -run=AWSAPIGatewayBase -timeout 120m
=== RUN   TestAccAWSAPIGatewayBasePath_basic
--- FAIL: TestAccAWSAPIGatewayBasePath_basic (19.88s)
       	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

       		* aws_api_gateway_domain_name.test: Error creating API Gateway Domain Name: BadRequestException: The domain name you provided is already associated with an existing CloudFront distribution.
       		Remove the domain name from the existing CloudFront distribution or use a different domain name.
       		If you own this domain name and are not using it on an existing CloudFront distribution, please contact support.
       			status code: 400, request id: 9******c-6852-11e6-9b0c-df6fa******2
```

I reckon it's because someone (you) created this domain in AWS and forgot to delete it.

apparentlymart replied:
That's actually rather awkward here, since API gateway seems to verify that the cert is valid for the domain.

Perhaps we have no option here but to talk to the Go crypto libraries directly in these tests and issue some fresh certs each time the tests run. Shouldn't be hard to do, but kinda annoying to have that much logic in a test. :grimacing:

We can at least keep the private key "hard-coded", which is the part that tends to be slowest due to all of the crunching needed to produce a good amount of randomness.

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping.go:
```diff
@@ -0,0 +1,130 @@
+	return &schema.Resource{
+		Create: resourceAwsApiGatewayBasePathMappingCreate,
+		Read:   resourceAwsApiGatewayBasePathMappingRead,
+		Delete: resourceAwsApiGatewayBasePathMappingDelete,
```
It seems to be possible to update certain fields of this resource:
http://docs.aws.amazon.com/cli/latest/reference/apigateway/update-base-path-mapping.html

but I'm totally happy to treat this PR as 1st iteration and eventually add `Update()` in a followup PR.

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping.go:
```diff
@@ -0,0 +1,130 @@
+
+		if err != nil {
+			if err, ok := err.(awserr.Error); ok && err.Code() != "BadRequestException" {
+				return &resource.RetryError{
```
Is there any reason why not use the handy helpers here instead? i.e. `resource.NonRetryableError(err)` & `resource.RetryableError(err)`

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping.go:
```diff
@@ -0,0 +1,130 @@
+		}
+
+		id := fmt.Sprintf("%s/%s", d.Get("domain_name").(string), d.Get("base_path").(string))
+		d.SetId(id)
```
I think it would make more sense to keep these 2 lines outside of the retry block, but it's really a nitpick! 😃

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_domain_name.go:
```diff
@@ -0,0 +1,181 @@
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NotFoundException" {
+			d.SetId("")
```
Like above - it would be nice to have `WARN` log message here IMO.

radeksimko commented:
@apparentlymart I left you some comments there, but vast majority of them are nitpicks. The only real biggie is one failing acceptance test due to uniqueness of domain names within AWS.

Otherwise this is looking pretty good.

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_domain_name_test.go:
```diff
@@ -0,0 +1,206 @@
+  certificate_chain = "%v"
+  certificate_name = "tf-acc-apigateway-domain-name"
+  certificate_private_key = "%v"
+  domain_name = "test-api-gateway-domain-cert0.com"
```
And this should be randomised as well.

apparentlymart commented:
Thanks for the feedback, @radeksimko. Agreed that randomizing that domain name seems unavoidable since it needs to be unique across the whole of AWS... will need to do some more work here to generate certs on the fly, and hope that API Gateway isn't *also* checking that the certificate chain leads back to some "real" CA that it trusts.... :worried:

apparentlymart commented:
Chatted a bit with @radeksimko elsewhere. He suggested that we should just replace the static, self-signed cert here with a wildcard one and then randomly generate a subdomain, which makes a lot of sense and should be easier than what I suggested.

@jarrodjackson pointed out that he made these certs using a Terraform config, which should be easy enough to adapt to make a wildcard cert: https://github.com/jarrodjackson/cert-generator

apparentlymart commented:
Using a wildcard cert seems to have worked to address the hostname conflict problem. I'll fix up the rest of this in a little while.

apparentlymart commented:
@radeksimko I've addressed the remaining comments now, except the one about doing updates on the base path mapping resource. I agree we should do this, but I also agree that it'd be good to address that in a separate PR since I think this approach will work for 90% of situations and make API Gateway setup with Terraform more complete.

(The main situation I see where it could potentially cause problems is if I want to change the stage of a pre-existing API. As implemented here Terraform will delete the old mapping and then create a replacement, which is likely to cause a brief outage where no mapping is present. Doing an update in-place would presumably avoid that. But I don't expect people will be frequently changing the stage of a deployed API.)

radeksimko commented on builtin/providers/aws/resource_aws_api_gateway_base_path_mapping.go:
```diff
@@ -0,0 +1,128 @@
+	})
+
+	id := fmt.Sprintf("%s/%s", d.Get("domain_name").(string), d.Get("base_path").(string))
+	d.SetId(id)
```
I think the ID should be set only if there are no errors - i.e. below the condition, not above. Theoretically `Read()` will wipe the resource during next `plan`, but it would be cleaner if it didn't ever end up in the state.

apparentlymart replied:
Oh yes, of course. :palm_tree:

radeksimko commented:
Besides one last nitpick this LGTM.

apparentlymart commented:
Thanks, @radeksimko. I addressed that last comment so I'm going to merge once the tests have passed.

apparentlymart commented:
Thanks to @jarrodjackson and @ryansroberts for doing all the work here! :grinning:

OliverCole commented:
While I wait for this to land in a relase - one question for @apparentlymart - you were talking about generating a certificate on the fly? What if we have a fully valid one? Will we be able to use that?

apparentlymart commented:
@OliverCole that issue with certs was specific to Terraform's acceptance tests. For real-world use, you can use any certificate you have as long as Terraform has access the PEM-encoded cert, cert chain and private key.

API Gateway is unfortunately divergent from the "usual" AWS approach: normally we can register a certificate separately in IAM/ACM and then reference it from other services, but for API Gateway it's necessary to upload the cert to the API Gateway API and then it will create a cert in IAM on your behalf. We can't do anything about this different workflow, but the Terraform resources will give you the same capabilities as you get when registering domains and path mappings via the UI.

OliverCole commented:
@apparentlymart Great, thanks for clarifying, and thanks for working on it!
